Debugging Menu Choices and Improving User Experience in Library Program | PORT > WIP

aCyberVoid opened the pull request:
This pull request focuses on debugging and improving the user experience of the Library program, especially concerning Menu Choices.

1. **Choice 1 and 3 Working but with Bugs**: While Choices 1 and 3 are operational, they are not flawless. Specifically, the bug arises when a user tries to see the books they have checked out before doing anything else with the menu. The program returns nothing in this case. However, if the user checks out a book first, the feature works as expected. This behavior needs to be debugged to provide a more consistent user experience.
2. **CSV File Update**: Successfully implemented the functionality to update a CSV file to reflect books checked out based on the user's name. While this is a big step forward, additional refinements are needed.
3. **Choices 2 and 4 (WIP)**: These menu choices are still a work in progress and need further development to be fully operational.
4. **UI/UX Improvement**: Current print statements output everything line by line, which makes the interface look cluttered. The plan is to incorporate better formatting, likely with new lines or other delineators, to break the information into more readable sections.
5. **Test File**: Created a separate test file to iteratively test different selection options and uncover additional errors that need resolving.

This pull request aims to address these issues and improve the overall usability of the program.

courtney-sims commented on Projects/MAC_Library_2/main.py:
```diff
@@ -0,0 +1,26 @@
+
+# Import required modules
+import csv
+from books import Book, RestrictedBook
```
Style-wise, you'll want an empty line between lines 4 and 5 https://peps.python.org/pep-0008/#imports

aCyberVoid replied:
🛠️ Updated!

courtney-sims commented on Projects/MAC_Library_2/library.py:
```diff
@@ -1,5 +1,105 @@
+import csv
+from books import Book, RestrictedBook
```
same thing here about the newline

aCyberVoid replied:
🛠️ Updated!

courtney-sims commented on Projects/MAC_Library_2/library.py:
```diff
@@ -1,5 +1,105 @@
+            next(books_reader)  # Skip the header row
+            for row in books_reader:
+                # print(f"\nRow data: {row}") - TESTING to view books.csv data
+                if len(row) == 5:
```
Is there a way to use the header of the csv file to parse this info instead?

courtney-sims commented on Projects/MAC_Library_2/library.py:
```diff
@@ -1,5 +1,105 @@
+                    writer.writerow([book.title, book.author, 'yes', ','.join(book.restricted_class), book.checked_out_by])
+                else:
+                    writer.writerow([book.title, book.author, 'no', '', book.checked_out_by])
+    def display_checked_out_books(self):
```
What is your desired behavior here when the user has no books checked out?

aCyberVoid replied:
If they do not have any books checked out, it lets them know and returns to the menu.

courtney-sims commented on Projects/MAC_Library_2/library.py:
```diff
@@ -1,5 +1,105 @@
+            return
+
+        # Check if the book is available
+        self.books[book_title].checked_out_by = self.current_student
```
This doesn't seem like it's doing the check mentioned in the comment above

aCyberVoid replied:
Removed the code.

courtney-sims commented on Projects/MAC_Library_2/library.py:
```diff
@@ -1,5 +1,105 @@
+        self.students[self.current_student].books_checked_out[book_title] = 0 
+
+        # Update the books.csv file
+        self.update_books_csv()
```
This is rewriting the entire books csv each time rather than just the item changed. Why did you choose that approach?

aCyberVoid commented:
# Updates

* Removed unused import: `from students import load_students`
* Removed **load_books** from books.py as its within library.py
* Wrapped the code inside a **try-except** block to catch any file-related errors as it interacts with a .csv file
* Added a check to see if book is available before allowing student to check out a book in the **check_out_book**
* Added a check for **return_book** function to see if the book is already checked in before updating the **check_out_by** attribute. This prevents unnecessary changes to csv file
* Added checks to **return_book**
 * Whether the book exists in the library.
 * Whether the book is already returned.
 * Whether the book is checked out by the current student.

Biggest Tip:
* Focused on each menu choice individually
* Thought of ways to test out parts of the code

aCyberVoid commented:
# Updates
- Created a test file where it uses the student name as "_Snowball_"
- Rearranged some lines of code to ensure functions were running in an order that allow proper file updating to books.csv
- Verified that option "_See checked out books_" will work PRIOR to needing to check out a book.
- Added additional print statements in **check_out_book(self)** for debugging 

# Issues
- Noticed most of my program is in library.py and does not interact with books.py yet. Will spend time ensuring I have code in the right places, and if so, maybe books.py is not needed. Less files.
- Students can check out any book, so need to add checks for **check_out_book** to prevent students from taking restricted books not in their restricted class.

aCyberVoid commented:
# Quick Update
### Changes
- Rearranged the debugging statements in `def updates_books_csv`. So far its only printing out non-restricted classes.
### To Do
- Need to complete choice 4. Goal is to have it default to page, and each time the student checks what page they are on, it increases by one.
- Move **updates_books_csv** to books.py and call its method in library.py. This can help reduce the lines of library.py and focus on the **library_menu** code and move other methods out that may not need to be there.
- Students are able to check out any books, despite restricted status. -- Continue to investigate
- When checking current page number, it does not default to 5 and it does not increase by 1 each time the student is checking. -- Continue to investigate

courtney-sims commented:
I think this is why it does not default to 5: https://github.com/aCyberVoid/PythonScrub/blob/ab8ba5cfb6a349a2acf1d03c12ebd3754d4d59c2/Projects/MAC_Library_2/library.py#L143

Where is the code that's meant to increment the page count?

courtney-sims commented:
And in terms of the restricted status, I don't see any code in check_out_books that looks for this
